File: scripts/generate-sitemap.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const rootDir = path.resolve(__dirname, '..');

// 1. Extract coding challenge slugs
const problemsFilePath = path.join(
  rootDir,
  'src/problem-engine/data/problems.ts'
);
const problemsContent = fs.readFileSync(problemsFilePath, 'utf-8');
const problemSlugRegex = /slug:\s*'([^']+)'/g;
let match;
const problemSlugs = [];
while ((match = problemSlugRegex.exec(problemsContent)) !== null) {
  problemSlugs.push(match[1]);
}

// 2. Extract learn curriculum lesson slugs
const curriculumFilePath = path.join(rootDir, 'src/learn/data/curriculum.ts');
const lessonSlugs = [];
if (fs.existsSync(curriculumFilePath)) {
  const curriculumContent = fs.readFileSync(curriculumFilePath, 'utf-8');
  const lessonGroupRegex = /lessonSlugs:\s*\[([\s\S]*?)\]/g;
  let groupMatch;
  while ((groupMatch = lessonGroupRegex.exec(curriculumContent)) !== null) {
    const itemRegex = /'([^']+)'/g;
    let itemMatch;
    while ((itemMatch = itemRegex.exec(groupMatch[1])) !== null) {
      lessonSlugs.push(itemMatch[1]);
    }
  }
}

const baseUrl = process.env.VITE_SITE_URL || 'https://runjs.in';
const lastMod = new Date().toISOString().split('T')[0] + 'T00:00:00+00:00';

const coreRoutes = [
  { path: '', changefreq: 'daily', priority: '1.0' },
  { path: '/learn', changefreq: 'daily', priority: '0.95' },
  { path: '/problems', changefreq: 'daily', priority: '0.9' },
  { path: '/js', changefreq: 'weekly', priority: '0.85' },
  { path: '/ts', changefreq: 'weekly', priority: '0.85' },
  { path: '/react', changefreq: 'weekly', priority: '0.85' },
  { path: '/html', changefreq: 'weekly', priority: '0.85' },
  { path: '/interview', changefreq: 'weekly', priority: '0.8' },
  { path: '/output-questions', changefreq: 'weekly', priority: '0.8' },
  { path: '/about', changefreq: 'monthly', priority: '0.7' },
  { path: '/privacy', changefreq: 'monthly', priority: '0.5' },
  { path: '/terms', changefreq: 'monthly', priority: '0.5' },
];

let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
xml += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n';

for (const route of coreRoutes) {
  xml += '  <url>\n';
  xml += `    <loc>${baseUrl}${route.path ? route.path : '/'}</loc>\n`;
  xml += `    <lastmod>${lastMod}</lastmod>\n`;
  xml += `    <changefreq>${route.changefreq}</changefreq>\n`;
  xml += `    <priority>${route.priority}</priority>\n`;
  xml += '  </url>\n';
}

for (const slug of lessonSlugs) {
  xml += '  <url>\n';
  xml += `    <loc>${baseUrl}/learn/${slug}</loc>\n`;
  xml += `    <lastmod>${lastMod}</lastmod>\n`;
  xml += `    <changefreq>weekly</changefreq>\n`;
  xml += `    <priority>0.8</priority>\n`;
  xml += '  </url>\n';
}

for (const slug of problemSlugs) {
  xml += '  <url>\n';
  xml += `    <loc>${baseUrl}/problems/${slug}</loc>\n`;
  xml += `    <lastmod>${lastMod}</lastmod>\n`;
  xml += `    <changefreq>monthly</changefreq>\n`;
  xml += `    <priority>0.8</priority>\n`;
  xml += '  </url>\n';
}

xml += '</urlset>\n';

const publicSitemapPath = path.join(rootDir, 'public/sitemap.xml');
fs.writeFileSync(publicSitemapPath, xml, 'utf-8');

const totalUrls = coreRoutes.length + lessonSlugs.length + problemSlugs.length;
console.log(
  `Generated sitemap.xml at ${publicSitemapPath} with ${totalUrls} URLs.`
);

File: scripts/generate_sitemap.py
#!/usr/bin/env python3
"""
Sitemap Generator for RunJS
Generates a comprehensive public/sitemap.xml with:
- Top-level static routes (Home, Learn, Problems, Playgrounds, About)
- All 175 JavaScript curriculum lessons
- All coding interview problems
"""

import os
import re
from datetime import datetime, timezone

BASE_URL = "https://runjs.rigial.com"
PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SITEMAP_PATH = os.path.join(PROJECT_ROOT, "public", "sitemap.xml")
ROBOTS_PATH = os.path.join(PROJECT_ROOT, "public", "robots.txt")

STATIC_ROUTES = [
    ("", 1.0, "daily"),
    ("learn", 0.95, "daily"),
    ("problems", 0.9, "daily"),
    ("js", 0.85, "weekly"),
    ("ts", 0.85, "weekly"),
    ("react", 0.85, "weekly"),
    ("html", 0.85, "weekly"),
    ("interview", 0.8, "weekly"),
    ("about", 0.6, "monthly"),
]

def extract_lesson_slugs():
    curriculum_path = os.path.join(PROJECT_ROOT, "src", "learn", "data", "curriculum.ts")
    slugs = []
    with open(curriculum_path, "r", encoding="utf-8") as f:
        content = f.read()
    blocks = re.findall(r'"lessonSlugs":\s*\[(.*?)\]', content, re.DOTALL)
    for b in blocks:
        for s in re.findall(r'"([a-z0-9\-]+)"', b):
            if s not in slugs:
                slugs.append(s)
    return slugs

def extract_problem_slugs():
    problems_path = os.path.join(PROJECT_ROOT, "src", "problem-engine", "data", "problems.ts")
    slugs = []
    with open(problems_path, "r", encoding="utf-8") as f:
        content = f.read()
    matches = re.findall(r"slug:\s*'([a-z0-9\-]+)'", content)
    for m in matches:
        if m not in slugs:
            slugs.append(m)
    return slugs

def generate():
    today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    urls = []

    # 1. Static Routes
    for path, priority, freq in STATIC_ROUTES:
        loc = f"{BASE_URL}/{path}".rstrip("/")
        urls.append(f"""  <url>
    <loc>{loc}</loc>
    <lastmod>{today}</lastmod>
    <changefreq>{freq}</changefreq>
    <priority>{priority:.2f}</priority>
  </url>""")

    # 2. Curriculum Lessons
    lesson_slugs = extract_lesson_slugs()
    print(f"Extracted {len(lesson_slugs)} curriculum lesson slugs.")
    for slug in lesson_slugs:
        loc = f"{BASE_URL}/learn/{slug}"
        urls.append(f"""  <url>
    <loc>{loc}</loc>
    <lastmod>{today}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.80</priority>
  </url>""")

    # 3. Problem Challenges
    problem_slugs = extract_problem_slugs()
    print(f"Extracted {len(problem_slugs)} problem challenge slugs.")
    for slug in problem_slugs:
        loc = f"{BASE_URL}/problems/{slug}"
        urls.append(f"""  <url>
    <loc>{loc}</loc>
    <lastmod>{today}</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.75</priority>
  </url>""")

    xml_content = f"""<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
        http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
{chr(10).join(urls)}
</urlset>
"""

    with open(SITEMAP_PATH, "w", encoding="utf-8") as f:
        f.write(xml_content)
    print(f"Wrote {len(urls)} URLs to {SITEMAP_PATH}")

    # Update robots.txt
    robots_content = f"""User-agent: *
Allow: /

# Sitemaps
Sitemap: {BASE_URL}/sitemap.xml
"""
    with open(ROBOTS_PATH, "w", encoding="utf-8") as f:
        f.write(robots_content)
    print(f"Updated {ROBOTS_PATH}")

if __name__ == "__main__":
    generate()
